### PR TITLE
[Fix] Apply weight decay in `SignSGD`

### LINKED ISSUE #517: `SignSGD` accepts `weight_decay` but never applies it
## Describe the bug

`SignSGD` accepts `weight_decay` and `weight_decouple`, validates them, and stores them in
`defaults`, but neither value is used anywhere in the step path.

- `sgd.py:429` validates `weight_decay`, and `sgd.py:437-438` store both in `defaults`.
- `_step_foreach` (`sgd.py:470`) and `_step_per_param` (`sgd.py:487`) never call
  `apply_weight_decay_foreach` or `apply_weight_decay`.
- The base helper at `base/optimizer.py:135` is not applied automatically. `SGDW` calls it
  explicitly at `sgd.py:205` and `sgd.py:242`.

So `SignSGD(params, lr=..., weight_decay=0.01)` runs without any weight decay and gives no warning.

I am filing this separately from #516 because fixing it changes results for anyone currently
passing a non-zero `weight_decay`, so it seemed better to ask than to send a patch. If you want the
fix, `Tiger` looks like the right template rather than `SGDW`: it is also a sign-based update, and
it applies decay after gradient maximisation and before the moment update (`tiger.py:92-98` and
`tiger.py:116-126`). Applying it after the buffer update instead leaves coupled decay outside the
`sign()` and it has no effect. I have a working patch and tests if that is useful.

## To Reproduce

* OS : macOS
* PyTorch version : 2.9.1
* Python version : 3.12
* pytorch-optimizer version : 3.10.1 (also reproduced on `main` at b452a0f)
* reproducible codes :

```python
import torch
from pytorch_optimizer import SignSGD

for weight_decouple in (True, False):
    for foreach in (False, True):
        p = torch.nn.Parameter(torch.full((2,), 2.0))
        opt = SignSGD([p], lr=0.1, momentum=0.9, weight_decay=0.5,
                      weight_decouple=weight_decouple, foreach=foreach)
        p.grad = torch.zeros(2)
        opt.step()
        print(f"weight_decouple={weight_decouple} foreach={foreach} -> p={p[0].item():.4f}")
```

Output:

```
weight_decouple=True foreach=False -> p=2.0000
weight_decouple=True foreach=True  -> p=2.0000
weight_decouple=False foreach=False -> p=2.0000
weight_decouple=False foreach=True  -> p=2.0000
```

With decoupled decay at `lr=0.1` and `weight_decay=0.5`, the parameter should move from `2.0` to
`1.9`. It does not move at all.

### PR DESCRIPTION
**Pull request by 11NOel11**
## Problem (Why?)

close #517

`SignSGD` validates `weight_decay` and stores it with `weight_decouple` in `defaults`, but neither is
used in either step path, so both are silently ignored.

## Solution (What/How?)

- [x] apply weight decay in `_step_foreach` and `_step_per_param`

Both calls go after gradient maximisation and before the momentum buffer update, following `Tiger`.
`SGDW` applies it after the buffer update instead, which is fine there because `grad = buf` aliases the
buffer; in `SignSGD` the buffer and the gradient stay separate, so decay applied after the lerp would be
dropped for the coupled case.

`fixed_decay=False` is hardcoded as in `SGDW`, since `SignSGD` has no `fixed_decay` argument. Happy to
add one if you would rather have it configurable.

## Other changes (bug fixes, small refactors)

N/A

## Notes

This changes results for anyone currently passing a non-zero `weight_decay`, since it did nothing
before. The default path is unaffected: at `weight_decay=0.0` the parameters are bit-identical to `main`
across momentum, `weight_decouple` and `foreach` settings.

One thing I noticed while testing, not caused by this change: the two step paths can disagree when the
momentum buffer lands near zero, since `_step_per_param` uses `buf.mul_().add_()` and `_step_foreach` uses
`_foreach_lerp_` - algebraically equal, different rounding, and `sign()` turns `1e-17` into a full `lr`
step. It reproduces on `main` at `weight_decay=0.0`. Happy to file it separately.

Three tests added, each parametrised over both `foreach` paths. With `p=2.0`, zero gradient, `lr=0.1`,
`momentum=0.9`, `weight_decay=0.2`: decoupled gives `2.0 * (1 - 0.2 * 0.1) = 1.96`, coupled gives
`grad = 0.2 * 2.0 = 0.4` so the update is `-lr * sign(0.04) = -0.1` and `p = 1.9`. The four decay tests
fail on `main` and pass with this change; the `weight_decay=0.0` test passes on both.

## Checklist

- [x] Make sure to run `just format` before commit
- [x] My code adheres to the style guidelines of this project (`just check` shows no errors)
- [x] Both new and existing unit tests pass successfully on my local environment by running `just test`
- [ ] I have made the necessary changes to the documentation

`ruff check` is clean on both files; I did not run `pyright`. Test suite: 2488 passed before, 2494 after,
with the same two pre-existing failures (`test_complex_optimizers[FTRL_...]` and `test_parse_version`).

### CODE REVIEW

**Review by kozistr — APPROVED** [review]
looks good to me! thanks for your contribution :)
